packages/client/src/app/ui/Chat/Chat.tsx

```diff
@@ -110,30 +190,112 @@ export function Chat() {
     scrollIntoViewRef.current?.scrollIntoView();
   }, [blurInput, lastInteraction, now]);
 
+  const getPlayerName = useCallback(() => {
+    // We think the user is a spectator
+    if (!currentPlayer.player) {
+      return 'Spectator';
+    }
+
+    const currentPlayerData = playerData.find((pd: any) => pd.player === currentPlayer.player);
+
+    if (currentPlayerData) {
+      return currentPlayerData.name;
+    }
+
+    // default to 'Spectator' if we don't know
+    return 'Spectator';
+  }, [currentPlayer, playerData]);
+
+  const getEncryptedTextContent = ({ player, text }: {player: string, text: string}) => {
+    const privKey = getBurnerWallet();
+    const recipientKey = players?.find(c => c.player === player)?.key;
+
+    if (!recipientKey) return null;
+
+    const sharedSecret = secp256k1.getSharedSecret(privKey.slice(2), recipientKey.slice(2));
+
+    const aesKey = sha256(sharedSecret);
+    const nonce = randomBytes(24);
+    const aesAlgo = gcm(aesKey, nonce);
+    const ciphertext = aesAlgo.encrypt(utf8ToBytes(text));
+
+    return {
+      ciphertext: bytesToHex(ciphertext),
+      nonce: bytesToHex(nonce)
+    };
+  }
+
+  const getDecryptedTextContent = ({player, ciphertext, nonce}: {player: string, ciphertext: string, nonce: string}) => {
+
+    const privKey = getBurnerWallet();
+    const recipientKey = players?.find(c => c.player === player)?.key;
```

Comment:
Will this break if a player registers two registration keys?

Reply:
In theory the `useEffect` block before this should prevent this, which only calls the registration function if the player's entry isn't found yet. 

But also the public key being registered is deterministic off of the `sessionWallet`'s private key, which doesn't appear to change -- it gets created once and then returned if it exists already.

```
    const matchingPlayers = players.filter((player: Player) => {
      return (player.key === publicKey)
    });

    if (matchingPlayers.length > 0) {
      return
    } else {
      registerEncryptionKey();
    }
```